src/Protocol/Diaspora/Repository/DiasporaContact.php

```diff
@@ -0,0 +1,286 @@
+	{
+		if ($update !== self::ALWAYS_UPDATE) {
+			try {
+				$dcontact = $this->selectOneByUriId(ItemURI::getIdByURI($uri));
+				if ($update === self::NEVER_UPDATE) {
+					return $dcontact;
+				}
+			} catch (HTTPException\NotFoundException $e) {
+				if ($update === self::NEVER_UPDATE) {
+					throw $e;
+				}
+
+				// This is necessary for Contact::getByURL in case the base contact record doesn't need probing,
+				// but we still need the result of a probe to create the missing diaspora-contact record.
+				$update = self::ALWAYS_UPDATE;
+			}
+		}
+
+		$contact = Contact::getByURL($uri, $update, ['uri-id']);
+		if (empty($contact['uri-id'])) {
+			throw new HTTPException\NotFoundException('Diaspora profile with URI ' . $uri . ' not found');
+		}
+
+		return self::selectOneByUriId($contact['uri-id']);
+	}
+
+	/**
+	 * Update or create a diaspora-contact entry via a probe array
+	 *
+	 * @param array $data Probe array
+	 * @return Entity\DiasporaContact
+	 * @throws \Exception
+	 */
+	public function updateFromProbeArray(array $data): Entity\DiasporaContact
+	{
+		$uriId = ItemURI::insert(['uri' => $data['url'], 'guid' => $data['guid']]);
+
+		$contact   = Contact::getByUriId($uriId, ['id', 'created']);
+		$apcontact = APContact::getByURL($data['url'], false);
+		if (!empty($apcontact)) {
+			$interacting_count = $apcontact['followers_count'];
+			$interacted_count  = $apcontact['following_count'];
+			$post_count        = $apcontact['statuses_count'];
+		} elseif (!empty($contact['id'])) {
+			$last_interaction = DateTimeFormat::utc('now - 180 days');
+
+			$interacting_count = $this->db->count('contact-relation', ["`relation-cid` = ? AND NOT `follows` AND `last-interaction` > ?", $contact['id'], $last_interaction]);
+			$interacted_count  = $this->db->count('contact-relation', ["`cid` = ? AND NOT `follows` AND `last-interaction` > ?", $contact['id'], $last_interaction]);
+			$post_count        = $this->db->count('post', ['author-id' => $contact['id'], 'gravity' => [Item::GRAVITY_PARENT, Item::GRAVITY_COMMENT]]);
+		}
+
+		$DiasporaContact = $this->factory->createfromProbeData(
+			$data,
+			$uriId,
+			new \DateTime($contact['created'] ?? 'now', new \DateTimeZone('UTC')),
+			$interacting_count ?? 0,
+			$interacted_count ?? 0,
+			$post_count ?? 0
+		);
+
+		$DiasporaContact = $this->save($DiasporaContact);
+
+		$this->logger->info('Updated diaspora-contact', ['url' => (string) $DiasporaContact->url, 'callstack' => System::callstack(20)]);
+
+		return $DiasporaContact;
+	}
+
+	/**
+	 * get a url (scheme://domain.tld/u/user) from a given contact guid
+	 *
+	 * @param mixed $guid Hexadecimal string guid
+	 *
+	 * @return string the contact url or null
+	 * @throws \Exception
+	 */
+	public function getUrlByGuid(string $guid): ?string
+	{
+		try {
+			$DiasporaContact = $this->selectOne(['guid' => $guid]);
+			return (string)$DiasporaContact->url;
```

Comment:
This seems a little bit inperformant, selecting the whole entity  just for one field
--> shouldn't we use here a direct table call instead?
We're already in the repository, so it won't break the pattern doing so

Reply:
Good point, sounds good.

Reply:
Simplified.